libyara/parser.c

```diff
@@ -228,6 +228,73 @@ int yr_parser_emit_pushes_for_strings(
   return ERROR_SUCCESS;
 }
 
+int yr_parser_emit_pushes_for_rules(
+    yyscan_t yyscanner,
+    const char* identifier)
+{
+  YR_COMPILER* compiler = yyget_extra(yyscanner);
+
+  // Make sure the compiler is parsing a rule
+  assert(compiler->current_rule_idx != UINT32_MAX);
+
+  YR_RULE* rule;
+
+  const char* rule_identifier;
+  const char* target_identifier;
+
+  int matching = 0;
+
+  YR_NAMESPACE* ns = (YR_NAMESPACE*) yr_arena_get_ptr(
+      compiler->arena,
+      YR_NAMESPACES_TABLE,
+      compiler->current_namespace_idx * sizeof(struct YR_NAMESPACE));
+
+  // Can't use yr_rules_foreach here as that requires the rules to have been
+  // finalized (inserting a NULL rule at the end). This is done when
+  // yr_compiler_get_rules() is called, which also inserts a HALT instruction
+  // into the current position in the code arena. Obviously we aren't done
+  // compiling the rules yet so inserting a HALT is a bad idea. To deal with
+  // this I'm manually walking all the currently compiled rules (up to the
+  // current rule index) and comparing identifiers to see if it is one we should
+  // use.
+  uint32_t rule_idx;
+  for (rule_idx = 0; rule_idx <= compiler->current_rule_idx; rule_idx++)
+  {
+    rule = yr_arena_get_ptr(
```

Comment:
Actually you don't need to call `yr_arena_get_ptr` on each loop iteration. You can simply use `yr_arena_get_ptr` for getting the pointer to the first rule (rule_idx = 0) and you can iterate up to `compiler->current_rule_idx` advancing the pointer `sizeof(YR_RULE)`. All the `YR_RULE` structs are guaranteed to be contiguous in memory. Calling `yr_arena_get_ptr` is required only after you write something in the `YR_RULES_TABLE` buffer, because the write may need to reallocate the buffer to somewhere else and any pointer you hold gets invalid. But as soon as you don't write a new rule you are safe, and can iterate over the existing ones without calling `yr_arena_get_ptr` each time.